### ISSUE TITLE
Certification check for SI is costly/incorrect

### ISSUE THREAD

**Issue by tcrain** [SyncFree/antidote#130]
Current implementation requires storing a history of all updates and all commit records in a cache.  Most protocols avoid this by having a integer per key of the last time it was written, or if this is too costly, a hash table of integers (this can cause false aborts though).

Also the current implementation seems incorrect.  A transaction should abort if what it is updating has a newer version at the DC than what is given by the transaction's snapshot.  Right now it only takes into account updates that from the local DC and it uses the TxId for ordering, but the TxId doesn't give commit order.

**Comment by tcrain** [comment]
Actually considering the part about aborting due to updates being received from other DCs is debatable.  On one hand you could say that the local transaction shouldn't abort because the update from the other DC could have just as easily arrived after the local transaction committed, thus having the same affect, especially if were are only considering certain mergeable objects.  But it is not quite the same because first it allows concurrent readers to observe different orders of the history that wouldn't be observable if the local transaction aborted and also it might matter for certain objects where the order of operations is important, which might be interesting for certain types of objects that could have "home DCs" where operations that originate at the home DC are different or are treated differently than operations that do not.  Possibly worth thinking about.

**Comment by marsleezm** [comment]
Hi, what you are considering seems to relate to Parallel Snapshot Isolation(http://news.cs.nyu.edu/~jinyang/pub/walter-sosp11.pdf). What they do is when reading, a transaction only reads the latest result from local DC but updates to each key still have establish a total order among DCs, which can be achieved by certifying writes in all DCs or having a primary to serialize updates. In practice, they avoid serializing updates by only using CRDT(called c-set).

I agree that the current certification has weird semantics, because it only serializes operations in a single DC but not among DC. While on the other hand, since we are only using CRDTs, we only need it to tell if a transaction can commit or not (e.g. for bounded counter), rather than establishing a total order for writes.

**Comment by tcrain** [comment]
I agree it has a similar air to PSI, but it is not quite the same because here any DC is accepting the updates, thus not admitting a total order of writes.

The current version (when fixed to use commit order instead of start time) is not quite snapshot isolation at a single DC either because an update can arrive from a different DC that invalidates your snapshot locally, but you will still commit.  But like I mentioned before this might not matter for many workloads because for most cases you eventually have the same result (but see different states in-between).

Like you say, validating only on local updates could be the correct way to implement this, especially if that is only what is needed for interesting objects like the bounded counter.  But I still think it is worth thinking about validating on all updates because there might be workloads that could take use of this (maybe for future versions of antidote).

In any case we have to be precise about what guarantees are provided so that people writing programs on top will know what to expect.

**Comment by marc-shapiro** [comment]
I'm worried that if you allow internal parallelism inside a DC you might violate HB, specifically program-order. 

                        Marc

**Comment by tcrain** [comment]
I should write an example also:

Consider you have 2 DCs, DC_A and DC_B and an object x with initial value v_init

Transaction T_A starts on DC_A
T_A reads x and sees v_init
Transaction T_B starts on DC_B
T_B updates x setting its value to v_B
T_B commits at DC_B
T_B is propagated to DC_A where it is applied
T_A updates x setting its value to v_A
T_A tries to commit at DC_A

Does T_A abort or commit?

**Comment by preguica** [comment]
In your example, if x is a crdt, I would say that T_A should commit.


Is there a specification of the protocols implemented in Antidote
somewhere? It is very hard to follow these threads with a spec of the
algorithms. (the same for the  Single receiver per DC for interDC traffic
thread and others)

**Comment by deepthidevaki** [comment]
On Thu, Jun 11, 2015 at 11:39 PM, tcrain <notifications@github.com> wrote:

> I should write an example also:
>
> Consider you have 2 DCs, DC_A and DC_B and an object x with initial value
> v_init
>
> Transaction T_A starts on DC_A
> T_A reads x and sees v_init
> Transaction T_B starts on DC_B
> T_B updates x setting its value to v_B
> T_B commits at DC_B
> T_B is propagated to DC_A where it is applied
> T_A updates x setting its value to v_A
> T_A tries to commit at DC_A
>
> Does T_A abort or commit?
>
T_A commits in this case. It aborts only if another transaction commits at
DC_A (which updates x).


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SyncFree/antidote/issues/130#issuecomment-111284253>.
>



-- 
------------------------------------------------------------------------------------------------------------------------------
“That which a man willingly shares, he keeps. That which he selfishly
keeps, he loses.”

**Comment by tcrain** [comment]
Sorry I should have written the issue report better from the start.

Currently Antidote supports some form of strong consistency per DC for local transactions.  This is implemented similar to Clock-SI where when a transaction wants to commit at its local DC, first it checks if its writes have been overwritten since the start time of the transaction.  This is enabled by default.  As far as I know there is no specification for this strong consistency part of Antidote, having a specification would help.

In the current implementation the validation of a transaction's writes only concerns values overwritten by local transactions, so in the example I gave above T_A will commit.  If you keep the same order of operations, but have T_B be a transaction that starts and executes at DC_A, then T_A will abort.  The question is then, why is this done this way?  I think there is no obvious answer, which is why I brought up the point.

Committing T_A makes sense if that provides the level on consistency needed for bounded counters, and that is the strongest level or consistency we want to support.

On the other hand committing T_A can make it confusing for a programmer as he/she can see that with the same order of operations sometimes a transaction will abort sometimes and sometimes it will commit (so he/she has to notice the origin of the operation to be sure).  Furthermore there might be applications that could use the stronger level of consistency provided by aborting T_B, I can given an example if there is interest.

**Comment by aletomsic** [comment]
> On Jun 12, 2015, at 11:26, tcrain <notifications@github.com> wrote:
> 
> Sorry I should have written the issue report better from the start.
> 
> Currently Antidote supports some form of strong consistency per DC for local transactions. This is implemented similar to Clock-SI where when a transaction wants to commit at its local DC, first it checks if its writes have been overwritten since the start time of the transaction. This is enabled by default. As far as I know there is no specification for this strong consistency part of Antidote, having a specification would help.
> 
This comes from the initial implementation of clock-si. As we agreed, the certification check should be disabled by default.
> In the current implementation the validation of a transaction's writes only concerns values overwritten by local transactions, so in the example I gave above T_A will commit. If you keep the same order of operations, but have T_B be a transaction that starts and executes at DC_A, then T_A will abort. The question is then, why is this done this way? I think there is no obvious answer, which is why I brought up the point.
> 
It should not be done that way. The certification check should be disabled and a every transaction should commit regardless of its origin.
> Committing T_A makes sense if that provides the level on consistency needed for bounded counters, and that is the strongest level or consistency we want to support.
> 
Bounded counters are the reason why we didn’t totally remove the certification check. But we have not discussed the implications of their support.

**Comment by tcrain** [comment]
The main issue is fixed here:
https://github.com/SyncFree/antidote/pull/160
